src/test/java/com/google/cloud/spanner/pgadapter/python/PythonTransactionTests.java

```diff
@@ -0,0 +1,1095 @@
+  }
+
+  @Test
+  public void testAutocommitInTransactions() throws IOException, InterruptedException {
+    // tests autocommit settings using connection.autocommit variable
+    List<String> statements = new ArrayList<>();
+    String sql1 = "Select * from some_table";
+    String sql2 = "insert into some_table(col1, col2) values(value1, value2)";
+    String sql3 = "Select * from some_table3";
+    String sql4 = "insert into some_table3(col1, col2) values(value1, value2)";
+
+    // no commit request will be sent after this query because autocommit is false
+    statements.add("query");
+    statements.add(sql1);
+
+    // no commit request will be sent after this update because autocommit is false
+    statements.add("update");
+    statements.add(sql2);
+
+    // no commit request will be sent after this update because autocommit is false
+    statements.add("update");
+    statements.add(sql4);
+
+    // 1 commit request will be sent
+    statements.add("transaction");
+    statements.add("commit");
+
+    statements.add("transaction");
+    statements.add("set autocommit True");
+
+    // no commit request after this query although autocommit is true because it is a query sql
+    statements.add("query");
+    statements.add(sql3);
+
+    // 1 commit request will be sent after this update because autocommit is true
+    statements.add("update");
+    statements.add(sql2);
+
+    // 1 commit request will be sent after this update because autocommit is true
+    statements.add("update");
+    statements.add(sql4);
+
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql1), createResultSet(1, "abcd")));
+    mockSpanner.putStatementResult(StatementResult.update(Statement.of(sql2), 1));
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sql3), createResultSet(2, "pqrs")));
+    mockSpanner.putStatementResult(StatementResult.update(Statement.of(sql4), 2));
+
+    String expectedOutput = "(1, 'abcd')\n" + "1\n" + "2\n" + "(2, 'pqrs')\n" + "1\n" + "2\n";
+    String actualOutput = executeTransactions(pgServer.getLocalPort(), statements);
+    for (AbstractMessage request : mockSpanner.getRequests())
+      System.out.println(request.getClass());
+    Assert.assertEquals(expectedOutput, actualOutput);
+
+    Assert.assertEquals(3, mockSpanner.countRequestsOfType(CommitRequest.class));
+    Assert.assertEquals(0, mockSpanner.countRequestsOfType(RollbackRequest.class));
+    Assert.assertEquals(6, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+
+    List<String> sql = new ArrayList<>();
+
+    for (ExecuteSqlRequest request : mockSpanner.getRequestsOfType(ExecuteSqlRequest.class)) {
+      sql.add(request.getSql());
+    }
+
+    Assert.assertEquals(1, Collections.frequency(sql, sql1));
+    Assert.assertEquals(2, Collections.frequency(sql, sql2));
+    Assert.assertEquals(1, Collections.frequency(sql, sql3));
+    Assert.assertEquals(2, Collections.frequency(sql, sql4));
```

Comment:
This needs to verify the order of the execution of the different statements and commit requests. Now we don't really know which statements were auto-committed and which were part of a transaction.

Reply:
Done